Mathlib/Analysis/Calculus/FirstDerivativeTest.lean

```diff
@@ -0,0 +1,145 @@
+lemma localMax_of_deriv_Ioo {f : ℝ → ℝ} {a b c : ℝ} (g₀ : a < b) (g₁ : b < c)
+    (h : ContinuousAt f b)
+    (hd₀ : DifferentiableOn ℝ f (Set.Ioo a b))
+    (hd₁ : DifferentiableOn ℝ f (Set.Ioo b c))
+    (h₀ :  ∀ x ∈ Set.Ioo a b, 0 ≤ deriv f x)
+    (h₁ :  ∀ x ∈ Set.Ioo b c, deriv f x ≤ 0) : IsLocalMax f b :=
+  have hIoc : ContinuousOn f (Ioc a b) :=
+    Ioo_union_right g₀ ▸ ContinuousOn.union_continuousAt (isOpen_Ioo (a := a) (b := b))
+        (DifferentiableOn.continuousOn hd₀) (by simp_all)
+  have hIco : ContinuousOn f (Ico b c) :=
+    Ioo_union_left g₁ ▸ ContinuousOn.union_continuousAt (isOpen_Ioo (a := b) (b := c))
+        (DifferentiableOn.continuousOn hd₁) (by simp_all)
+  isLocalMax_of_mono_anti g₀ g₁
+    (monotoneOn_of_deriv_nonneg (convex_Ioc a b) hIoc (by simp_all) (by simp_all))
+    (antitoneOn_of_deriv_nonpos (convex_Ico b c) hIco (by simp_all) (by simp_all))
+
+
+/-- The First-Derivative Test from calculus, minima version. -/
+lemma localMin_of_deriv_Ioo {f : ℝ → ℝ} {a b c : ℝ} (h : ContinuousAt f b)
+    (g₀ : a < b) (g₁ : b < c)
+    (hd₀ : DifferentiableOn ℝ f (Set.Ioo a b)) (hd₁ : DifferentiableOn ℝ f (Set.Ioo b c))
+    (h₀ : ∀ x ∈ Set.Ioo a b, deriv f x ≤ 0)
+    (h₁ : ∀ x ∈ Set.Ioo b c, 0 ≤ deriv f x) : IsLocalMin f b := by
+    have := localMax_of_deriv_Ioo (f := -f) g₀ g₁
+      (by simp_all) (DifferentiableOn.neg hd₀) (DifferentiableOn.neg hd₁)
+      (fun x hx => deriv.neg (f := f) ▸ Left.nonneg_neg_iff.mpr <|h₀ x hx)
+      (fun x hx => deriv.neg (f := f) ▸ Left.neg_nonpos_iff.mpr <|h₁ x hx)
+    exact (neg_neg f) ▸ IsLocalMax.neg this
+
+/-- If `p` holds to the left of `a` then it holds in an open interval `(l, a)`. -/
+lemma Filter.Eventually.exists_lt_forall_Ioo {α : Type*} [TopologicalSpace α] [LinearOrder α]
+    [OrderTopology α] [NoMinOrder α] {a : α} {p : α → Prop} (h : ∀ᶠ x in 𝓝[<] a, p x) :
+    ∃ l < a, ∀ x ∈ Ioo l a, p x :=
+  mem_nhdsWithin_Iio_iff_exists_Ioo_subset.1 h
+
+/-- If `p` holds to the right of `a` then it holds in an open interval `(a, l)`. -/
+lemma Filter.Eventually.exists_gt_forall_Ioo {α : Type*} [TopologicalSpace α] [LinearOrder α]
+    [OrderTopology α] [NoMaxOrder α] {a : α} {p : α → Prop} (h : ∀ᶠ x in 𝓝[>] a, p x) :
+    ∃ l > a, ∀ x ∈ Ioo a l, p x :=
+  mem_nhdsWithin_Ioi_iff_exists_Ioo_subset.1 h
+
+
+/-- Monotonicity of open intervals under removal of `max` at a left endpoint. -/
+theorem mem_Ioo_of_mem_Ioo_max_left {R : Type*} [LinearOrder R]
+    {b u₀ v₀ x : R} (hx : x ∈ Ioo (max u₀ v₀) b) :
+    x ∈ Ioo u₀ b := by simp_all
+
+/-- Monotonicity of open intervals under removal of `min` at a right endpoint. -/
+theorem mem_Ioo_of_mem_Ioo_min_right {R : Type*} [LinearOrder R]
+    {b u₁ v₁ x : R} (hx : x ∈ Ioo b (min u₁ v₁)) :
+    x ∈ Ioo b u₁ := by simp_all
+
+/-- The interval inclusion `(a,b] \ {b} ⊆ (a,b)`. -/
+theorem mem_Ioo_of_mem_Ioc_of_ne {R : Type*} [LinearOrder R]
+    {b u₀ x : R} (hx : x ∈ Ioc u₀ b) (H : ¬x = b) :
+    x ∈ Ioo u₀ b := by
```

Comment:
Lemmas like this should probably not exist. You can use [Set.Ioc_diff_right](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Interval/Set/Basic.html#Set.Ioc_diff_right) to get this quickly.
Similar for the other ones, e.g. by using [Set.Ioo_subset_Ioo_left](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Interval/Set/Basic.html#Set.Ioo_subset_Ioo_left)

Reply:
OK, I removed that one and the similar one.
As for the ones with max and min (`mem_Ioo_of_mem_Ioo_max_left`), it seems that there is a tension with not-repeating-oneself in proofs, and with golfing. If I remove that those then the proof that uses them will be longer and also more repetitive... I'm not sure how to reconcile those concerns.